lib/connectors_async/job_watcher.rb

```diff
@@ -0,0 +1,73 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+require 'concurrent'
+
+require 'connectors_shared/logger'
+
+module ConnectorsAsync
+  class JobWatcher
+    IDLE_TIME = 30 # seconds
+
+    class AlreadyWatchingError < StandardError; end
+
+    def initialize(job_store:)
+      @job_store = job_store
+
+      @pool = Concurrent::ThreadPoolExecutor.new(
+        min_threads: 1,
+        max_threads: 1,
+        max_queue: 0,
+        idletime: IDLE_TIME + 1
+      )
+
+      @is_watching = Concurrent::AtomicBoolean.new(false)
+    end
+
+    def watch
+      raise AlreadyWatchingError.new('Already watching!') unless @is_watching.make_true
+      ConnectorsShared::Logger.info('Watching after jobs {•̃_•̃}')
+
+      @pool.post do
+        loop do
+          run!
+        rescue StandardError => e
+          ConnectorsShared::Logger.error("An error occurred while watching over jobs: #{e.message}")
```

Comment:
Let's use `ConnectorsShared::ExceptionTracking` instead, so that we can get the stack trace too